pkg/controller/tailoredprofile/tailoredprofile_controller.go

```diff
@@ -187,6 +275,320 @@ func (r *ReconcileTailoredProfile) Reconcile(ctx context.Context, request reconc
+					r.Eventf(v1alphaTp, corev1.EventTypeWarning, "TailoredProfileMigratedRule", "Removing migrated rule: %s from disableRules", rule.Name)
+				} else {
+					logger.Info("Migrated rule detected in disableRules", "rule", rule.Name)
+					r.Eventf(v1alphaTp, corev1.EventTypeWarning, "TailoredProfileMigratedRule", "Migrated rule detected: %s Please migrate it or remove it from the TailoredProfile", rule.Name)
+					ruleNeedToBeMigratedList = append(ruleNeedToBeMigratedList, rule.Name)
+					newRules = append(newRules, *rule)
+				}
+				continue
+			}
+			newRules = append(newRules, *rule)
+		}
+		if len(newRules) != len(v1alphaTp.Spec.DisableRules) {
+			v1alphaTpCP.Spec.DisableRules = newRules
+		}
+	}
+
+	// check if there are any enabled rules that are migrated
+	if len(v1alphaTp.Spec.EnableRules) > 0 {
+		var newRules []cmpv1alpha1.RuleReferenceSpec
+		for ri := range v1alphaTp.Spec.EnableRules {
+			rule := &v1alphaTp.Spec.EnableRules[ri]
+			if checkType, ok := migratedRules[rule.Name]; ok && checkType != string(profileType) {
+				if pruneOudated {
+					doContinue = false
+					logger.Info("Removing migrated rule from enableRules", "rule", rule.Name)
+					r.Eventf(v1alphaTp, corev1.EventTypeWarning, "TailoredProfileMigratedRule", "Removing migrated rule %s from enableRules", rule.Name)
+				} else {
+					logger.Info("Migrated rule detected in enableRules", "rule", rule.Name)
+					r.Eventf(v1alphaTp, corev1.EventTypeWarning, "TailoredProfileMigratedRule", "Migrated rule detected: %s Please migrate it or remove it from the TailoredProfile", rule.Name)
+					ruleNeedToBeMigratedList = append(ruleNeedToBeMigratedList, rule.Name)
+					newRules = append(newRules, *rule)
+				}
+				continue
+			}
+			newRules = append(newRules, *rule)
+		}
+
+		if len(newRules) != len(v1alphaTp.Spec.EnableRules) {
+			v1alphaTpCP.Spec.EnableRules = newRules
+		}
+	}
+
+	if v1alphaTp.Spec.Extends == "" && len(v1alphaTpCP.Spec.DisableRules) == 0 && len(v1alphaTpCP.Spec.EnableRules) == 0 {
+		errorMsg := "TailoredProfile does not have any rules left after removing migrated rules and it does not extend any profile"
+		v1alphaTpCP.Status.State = cmpv1alpha1.TailoredProfileStateError
+		v1alphaTpCP.Status.ErrorMessage = errorMsg
+		doContinue = false
+		logger.Info(errorMsg)
+		r.Eventf(v1alphaTp, corev1.EventTypeWarning, "TailoredProfileMigratedRule", errorMsg)
+	}
+
+	if !doContinue {
+		err = r.Client.Update(context.TODO(), v1alphaTpCP)
+		logger.Info("Updating TailoredProfile after handling migration")
+		if err != nil {
+			return false, nil, err
+		}
+	}
+	return doContinue, ruleNeedToBeMigratedList, nil
+}
+
+func isValidationRequired(tp *cmpv1alpha1.TailoredProfile) bool {
+	if tp.Spec.Extends != "" {
+		return tp.Spec.DisableRules != nil || tp.Spec.EnableRules != nil || tp.Spec.ManualRules != nil || tp.Spec.SetValues != nil
+	}
+	return tp.Spec.EnableRules != nil || tp.Spec.ManualRules != nil
```

Comment:
I think a TP that doesn't extend any profile can still `SetValues`, right?
So maybe its should also check whether there are `SetValues`?
```suggestion
	return tp.Spec.EnableRules != nil || tp.Spec.ManualRules != nil || tp.Spec.SetValues != nil
```

Reply:
that's a good catch fixing it now